### PR TITLE
Port test logic to Dart.

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
This gets rid of `dev/run_tests` and `packages/flutter_tools/bin/sky_test.dart`. Now to run the tests you just run `flutter test --debug`.

cc @abarth

File: packages/flutter_tools/lib/src/commands/flutter_command_runner.dart [hunk 1]
```diff
@@ -100,6 +115,17 @@ class FlutterCommandRunner extends CommandRunner {
...
   ArgResults _globalResults;
 
+  String get defaultFlutterRoot {
+    String script = Platform.script.toFilePath();
+    if (Platform.environment.containsKey(kFlutterRootEnvironmentVariableName))
+      return Platform.environment[kFlutterRootEnvironmentVariableName];
+    if (path.basename(script) == kSnapshotFileName)
+      return path.dirname(path.dirname(path.dirname(script)));
+    if (path.basename(script) == kFlutterToolsScriptFileName)
+      return path.dirname(path.dirname(path.dirname(path.dirname(script))));
```
Comment (abarth) [inline packages/flutter_tools/lib/src/commands/flutter_command_runner.dart]: Fancy

File: packages/flutter_tools/lib/src/commands/flutter_command_runner.dart [hunk 2]
```diff
@@ -108,39 +134,47 @@ class FlutterCommandRunner extends CommandRunner {
...
+      if (enginePath == null) {
+        String tryEnginePath(String enginePath) {
+          if (FileSystemEntity.isDirectorySync(path.join(enginePath, 'out')))
+            return enginePath;
+          return null;
+        }
+        enginePath = tryEnginePath(path.join(ArtifactStore.flutterRoot, '../$kFlutterEnginePackageName/src')) ??
+                     tryEnginePath(path.join(ArtifactStore.flutterRoot, '../engine/src')) ??
+                     tryEnginePath(path.join(ArtifactStore.flutterRoot, '../sky_engine/src')) ??
+                     tryEnginePath(path.join(ArtifactStore.flutterRoot, '../flutter_engine/src'));
```
Comment (abarth) [inline packages/flutter_tools/lib/src/commands/flutter_command_runner.dart]: I'm not sure we should have all these various fallbacks.  That's good for error tolerance, but it means folks will have a diversity of setups.  Instead, we should guide people towards the one true setup.  Maybe these alternatives are worth looking at to provide good error messages?

**Comment by abarth** [comment]
LGTM

Would you be willing to update CONTRIBUTING.md as well?  Eventually, I'd like to grow `flutter test` to be useful for end developers as well, but we can worry about that in future patches.

Thanks!

File: packages/flutter_tools/lib/src/commands/flutter_command_runner.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/flutter_command_runner.dart]: yeah, fair enough. just the first one?

Comment (abarth) [inline packages/flutter_tools/lib/src/commands/flutter_command_runner.dart]: The golden path in the rest of the docs is actually the second one (matching the repository name).  The first one has the name "sky" in it, which we'll presumably want to remove eventually.

Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/flutter_command_runner.dart]: roger.

**Comment by Hixie** [comment]
Updated as per comments. Will check in if Travis agrees. cc @abarth

**Pull request merged**